data_model/derive/src/lib.rs

```diff
@@ -14,216 +96,578 @@ use quote::{format_ident, quote};
+                    &self.origin_filter
+                }
+
+                /// Get `event_filter`
+                #[inline]
+                pub const fn event_filter(&self) -> &#fil_opt<#event_filter_ident> {
+                    &self.event_filter
+                }
+            }
+
+            impl Filter for #filter_ident {
+                type EventType = #imp_event;
+                fn matches(&self, event: &Self::EventType) -> bool {
+                    self.origin_filter.matches(event) && self.event_filter.matches(event)
+                }
+            }
+
+            #event_filter_and_impl
+        }
+        .into()
     }
-    .into()
-}
 
-fn process_event_variant(
-    variant_type: &syn::Type,
-    event_filter_variant_ident: &syn::Ident,
-    variants_with_id_fields: &mut Vec<syn::Ident>,
-    variants_with_event_fields: &mut Vec<syn::Ident>,
-) -> Option<syn::Ident> {
-    if let syn::Type::Path(path) = variant_type {
-        let var_ty_ident = &path.path.segments[0].ident;
-
-        var_ty_ident
-            .to_string()
-            .ends_with("Event")
-            .then(|| {
-                variants_with_event_fields.push(event_filter_variant_ident.clone());
-                format_ident!(
-                    "{}Filter",
-                    var_ty_ident
-                        .to_string()
-                        .strip_suffix("Event")
-                        .expect("Variant name should have suffix `Event`"),
-                )
-            })
-            .or_else(|| {
-                variants_with_id_fields.push(event_filter_variant_ident.clone());
-                None
-            })
-    } else {
-        None
+    /// Generates the event filter for the event. E.g. for `AccountEvent`, `AccountEventFilter`
+    /// and its `impl Filter` are generated.
+    fn impl_event_filter(event: &EventEnum) -> proc_macro2::TokenStream {
+        let EventEnum {
+            vis,
+            ident: event_ident,
+            generics,
+            ..
+        } = event;
+
+        let id_variants = event.generate_filter_variants_with_id_fields();
+        let event_variants = event.generate_filter_variants_with_event_fields();
+
+        let id_impls = event.generate_filter_impls_with_id_fields();
+        let event_impls = event.generate_filter_impls_with_event_fields();
+
+        let event_filter_ident = format_ident!("{}Filter", event_ident);
+        let import_path = quote! { crate::prelude };
+        let imp_event = quote! { #import_path::#event_ident };
+
+        quote! {
+            #[derive(
+                Clone,
+                PartialEq,
+                PartialOrd,
+                Ord,
+                Eq,
+                Debug,
+                Decode,
+                Encode,
+                Deserialize,
+                Serialize,
+                IntoSchema,
+                Hash,
+            )]
+            #[allow(clippy::enum_variant_names, missing_docs)]
```

Comment:
please don't allow `missing_docs` on public items. You can store documentation into `EventEnum`

Reply:
It's there to prevent lints screaming at no documentation for enum variants, such as `ByMetadataRemoved`. I believe they're quite self-descriptive, so it's hard to come up with useful docs, and the same allow attribute was in the original code as well.